docs/proposals/20221109-managed-external-etcd.md

```diff
@@ -0,0 +1,572 @@
+
+### Etcdadm cluster controller
+It accepts the etcd cluster configuration options from the user, such as the cluster size or version, and generates the objects required by the Etcdadm bootstrap provider for etcd cluster creation. It is responsible for managing the etcd cluster lifecycle.
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+### Abbreviations:
+- CAPI: Cluster API
+- KCP: KubeadmControlPlane
+- CRD: Custom Resource Definition
+
+## Summary
+
+This proposal aims to extend CAPI to support provisioning and managing of the lifecycle of external etcd clusters. With this, users can create CAPI clusters with the external etcd topology where the etcd cluster is also managed by CAPI.
+
+## Motivation
+
+- Cluster API supports the use of an external etcd cluster, by allowing users to provide their external etcd cluster's endpoints.
+So it would be good to add support for provisioning the etcd cluster too, so the user can have a single workflow for full management of a cluster using the external etcd topology. The user can offload the responsibility of managing an external etcd cluster to CAPI.
+    
+- External etcd topology decouples the control plane and etcd member. So if a control plane-only node fails, or if there's a memory leak in a component like kube-apiserver, it won't directly impact an etcd member.
+  
+- Etcd is resource intensive, so it is safer to have dedicated nodes for etcd, it could use more disk space, higher bandwidth. Having a separate etcd cluster for these reasons could ensure a more resilient HA setup.
+
+### Goals
+
+- Define a contract for adding pluggable etcd providers
+  - Define the responsibilities of an etcd provider, along with the required services and API types it should add.
+  - User should be able to create and manage a Kubernetes cluster that uses external etcd topology in a declarative way. The CAPI controllers will deploy the etcd cluster first and then use that as input to the workload cluster, without requiring any user intervention.
+  - There will be a 1:1 mapping between the external etcd cluster and the workload cluster.
+  - An etcd provider should support etcd clusters of any size (odd numbers only), including single member etcd clusters for testing purposes. But the documentation should specify the recommended cluster size for production use cases to range between 3 and 7.
+- Provide a first implementation of this pluggable etcd provider using [etcdadm](https://github.com/kubernetes-sigs/etcdadm) that integrates with the Kubeadm Control Plane provider.
+  - The etcdadm based provider will utilize the existing Machine objects to represent etcd members for convenience instead of adding a new machine type for etcd.
+  - Support the following etcd cluster management actions: scale up and scale down, etcd member replacement and etcd version upgrades.
```

Comment:
Should we add that we need some code reuse between this effort and KCP's current etcd capabilities? Ultimately we should try to maximize our shared code between the external etcd provider and stacked.